File: connector/src/types/trigger.ts
export interface TriggerBufferConfig {
  time_threshold_minutes: number;
  count_threshold: number;
  min_buffer_minutes: number;
}

export interface TriggerMatchConfig {
  confidence_threshold: number;
}

export interface TriggerScopeOption {
  label: string;
  value: string;
  description?: string | null;
}

export interface TriggerScopeField {
  key: string;
  label: string;
  type: string;
  multiple: boolean;
  required: boolean;
  help?: string | null;
  options: TriggerScopeOption[];
}

export interface TriggerConnectorOption {
  connector_id: string;
  connector_name: string;
  plugin_type: string;
  display_name: string;
  icon: string;
  adapter_kind: string;
  install_hint?: string | null;
  status: string;
  scope_fields: TriggerScopeField[];
}

export interface TriggerConnectorSummary {
  id: string;
  name: string;
  type: string;
  display_name: string;
  icon: string;
}

export interface TriggerStats {
  matched_events_last_24h: number;
  feature_request_count: number;
  open_buffer_events: number;
}

export interface Trigger {
  id: string;
  connector: TriggerConnectorSummary;
  plugin_type: string;
  natural_language_description: string;
  parsed_filter_criteria: Record<string, unknown>;
  scope: Record<string, unknown>;
  scope_summary: string;
  status: "active" | "paused" | "error" | string;
  buffer_config: TriggerBufferConfig;
  match_config: TriggerMatchConfig;
  stats: TriggerStats;
  last_event_at?: string | null;
  last_dispatch_at?: string | null;
  last_error?: string | null;
  created_at: string;
  updated_at: string;
}

export interface TriggerFeatureRequestLink {
  id: string;
  title: string;
}

export interface TriggerActivityEvent {
  id: string;
  external_id: string;
  match_score?: number | null;
  processing_status: string;
  content_text: string;
  source_label: string;
  author_name?: string | null;
  signal_id?: string | null;
  created_at: string;
  processed_at?: string | null;
  feature_requests: TriggerFeatureRequestLink[];
}

export interface TriggerBuffer {
  id: string;
  event_count: number;
  status: string;
  buffer_started_at: string;
  last_event_at?: string | null;
  dispatched_at?: string | null;
  completed_at?: string | null;
  synthesis_run_id?: string | null;
  feature_request_ids: string[];
  error?: string | null;
}

export interface TriggerDetail {
  trigger: Trigger;
  recent_events: TriggerActivityEvent[];
  recent_buffers: TriggerBuffer[];
}

export interface TriggerPayload {
  connector_id: string;
  natural_language_description: string;
  scope: Record<string, unknown>;
  buffer_config: TriggerBufferConfig;
  match_config: TriggerMatchConfig;
  status?: string;
}

export interface TriggerUpdatePayload {
  natural_language_description?: string;
  scope?: Record<string, unknown>;
  buffer_config?: TriggerBufferConfig;
  match_config?: TriggerMatchConfig;
  status?: string;
}

File: synthesis/src/schemas/trigger.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class TriggerBufferConfig(BaseModel):
    time_threshold_minutes: int = 60
    count_threshold: int = 10
    min_buffer_minutes: int = 5


class TriggerMatchConfig(BaseModel):
    confidence_threshold: float = 0.7


class TriggerCreate(BaseModel):
    connector_id: UUID
    natural_language_description: str = Field(min_length=8, max_length=4000)
    scope: dict = Field(default_factory=dict)
    buffer_config: TriggerBufferConfig = Field(default_factory=TriggerBufferConfig)
    match_config: TriggerMatchConfig = Field(default_factory=TriggerMatchConfig)
    status: str = Field(default="active")


class TriggerUpdate(BaseModel):
    natural_language_description: str | None = Field(default=None, min_length=8, max_length=4000)
    scope: dict | None = None
    buffer_config: TriggerBufferConfig | None = None
    match_config: TriggerMatchConfig | None = None
    status: str | None = None


class TriggerScopeOption(BaseModel):
    label: str
    value: str
    description: str | None = None


class TriggerScopeField(BaseModel):
    key: str
    label: str
    type: str
    multiple: bool = False
    required: bool = False
    help: str | None = None
    options: list[TriggerScopeOption] = Field(default_factory=list)


class TriggerConnectorOption(BaseModel):
    connector_id: UUID
    connector_name: str
    plugin_type: str
    display_name: str
    icon: str
    adapter_kind: str
    install_hint: str | None = None
    status: str
    scope_fields: list[TriggerScopeField] = Field(default_factory=list)


class TriggerConnectorSummary(BaseModel):
    id: UUID
    name: str
    type: str
    display_name: str
    icon: str


class TriggerStats(BaseModel):
    matched_events_last_24h: int = 0
    feature_request_count: int = 0
    open_buffer_events: int = 0


class TriggerRead(BaseModel):
    id: UUID
    connector: TriggerConnectorSummary
    plugin_type: str
    natural_language_description: str
    parsed_filter_criteria: dict = Field(default_factory=dict)
    scope: dict = Field(default_factory=dict)
    scope_summary: str
    status: str
    buffer_config: TriggerBufferConfig = Field(default_factory=TriggerBufferConfig)
    match_config: TriggerMatchConfig = Field(default_factory=TriggerMatchConfig)
    stats: TriggerStats = Field(default_factory=TriggerStats)
    last_event_at: datetime | None = None
    last_dispatch_at: datetime | None = None
    last_error: str | None = None
    created_at: datetime
    updated_at: datetime


class TriggerFeatureRequestLink(BaseModel):
    id: UUID
    title: str


class TriggerActivityEvent(BaseModel):
    id: UUID
    external_id: str
    match_score: float | None = None
    processing_status: str
    content_text: str = ""
    source_label: str = ""
    author_name: str | None = None
    signal_id: UUID | None = None
    created_at: datetime
    processed_at: datetime | None = None
    feature_requests: list[TriggerFeatureRequestLink] = Field(default_factory=list)


class TriggerBufferRead(BaseModel):
    id: UUID
    event_count: int
    status: str
    buffer_started_at: datetime
    last_event_at: datetime | None = None
    dispatched_at: datetime | None = None
    completed_at: datetime | None = None
    synthesis_run_id: UUID | None = None
    feature_request_ids: list[str] = Field(default_factory=list)
    error: str | None = None


class TriggerDetail(BaseModel):
    trigger: TriggerRead
    recent_events: list[TriggerActivityEvent] = Field(default_factory=list)
    recent_buffers: list[TriggerBufferRead] = Field(default_factory=list)
